src/EmbedPreview.tsx

```diff
@@ -0,0 +1,532 @@
+          },
+          entry: (ctx) => {
+            console.debug(
+              'SENDING POSTMESSAGE',
+              ctx.params,
+              ctx.embedUrl.replace('/viz', ''),
+            );
+            ctx.iframe?.contentWindow?.postMessage(
+              {
+                type: 'EMBED_PARAMS_CHANGED',
+                url: ctx.embedUrl.replace('/viz', ''),
+              },
+              '*',
+            );
+          },
+        },
+        loading: {
+          tags: 'preview_loading',
+          on: {
+            IFRAME_LOADED: 'loaded',
+            IFRAME_ERROR: 'error',
+          },
+        },
+        loaded: {
+          entry: embedPreviewModel.assign({ loaded: true }),
+        },
+        error: {
+          tags: 'preview_error',
+          on: {
+            RETRY: 'loading',
+          },
+        },
+      },
+    },
+  },
+});
+
+const useEmbedCodeClipboard = () => {
+  const [value, setValue] = useState('');
+  const { onCopy, hasCopied } = useClipboard(value);
+
+  return {
+    copy: onCopy,
+    setCopyText: setValue,
+    isCopied: hasCopied,
+  };
+};
+
+const EmbedPreviewContent: React.FC = () => {
+  const router = useRouter();
+  const form = useRef<HTMLFormElement>(null!);
+  const {
+    copy: copyEmbedCode,
+    isCopied,
+    setCopyText,
+  } = useEmbedCodeClipboard();
+  const iframe = useRef<HTMLIFrameElement>(null!);
+  const [previewState, sendPreviewEvent, service] = useMachine(
+    embedPreviewMachine.withConfig({
+      actions: {
+        saveParams: assign({
+          params: (_, e) => {
+            console.debug('saveParams', e.params);
+            return (e as any).params;
+          },
+        }),
+        updateEmbedCopy: (ctx) => {
+          setCopyText(ctx.embedCode);
+        },
+        makeEmbedUrlAndCode: assign((ctx) => {
+          const url = makeEmbedUrl(
+            router.query.sourceFileId as string,
+            ctx.params,
+          );
+
+          return {
+            embedUrl: url,
+            embedCode: getEmbedCodeFromUrl(url),
+            params: ctx.params,
+          };
+        }),
+        makePreviewUrl: assign((ctx) => {
+          const url = makeEmbedUrl(
+            router.query.sourceFileId as string,
+            ctx.params,
+          );
+          return {
+            previewUrl: url,
+          };
+        }),
+      },
+    }),
+  );
+
+  useEffect(() => {
+    // service.onEvent((evt) => console.debug(evt.type));
+    service.onTransition((state, ctx) => {
+      console.debug(state.event, state.actions);
+    });
+  }, []);
+
+  useEffect(() => {
+    const formRef = form.current;
+    sendPreviewEvent({ type: 'IFRAME_READY', iframe: iframe.current });
+    return () => {
+      formRef.reset();
+    };
+  }, []);
```

Comment:
What does this do? Reminds me a bit of my legacy code talk 😂 

Could you add a comment? Or wrap in a custom hook?

Reply:
Hard to wrap in a custom hook. it does 2 things:
1. On mount, signal that iframe ref is ready
2. Reset the form when leaving the embed preview (unmount)